### PR TITLE
Search: add class to retrieve search product pricing

### LINKED ISSUE #23234: S2.1 Pricing API integration

### PR DESCRIPTION
**Pull request by kangzj**
Fixes #23234 

#### Changes proposed in this Pull Request:
Add a class to fetch search products from WPCOM and provides functionality to get tier information. REST API and Jetpack client changes will be proposed in a followup PR.

Note:

- Tests will pass but will need `D78501` to be functional
- <strike>Need a rebase after #23809 is merged.</strike>

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Ensure PHP tests pass `composer test-php`

### CODE REVIEW

File: projects/packages/search/src/class-search-products.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	 * @param int $record_count The number of record to estimate the tier.
+	 */
+	public function get_site_tier( $record_count = 0 ) {
+		$record_count = $record_count > 0 ? $record_count : Stats::estimate_count();
+		$product      = static::get_promoted_product();
+		if ( ! $record_count || ! isset( $product['price_tier_list'] ) ) {
+			return static::DEFAULT_TIER_INFO;
+		}
+		$price_tier_list = $product['price_tier_list'];
+		array_multisort( array_column( $price_tier_list, 'maximum_units' ), SOTR_ASC, $price_tier_list );
```
Comment (bluefuton) [inline projects/packages/search/src/class-search-products.php:48]:
```suggestion
		array_multisort( array_column( $price_tier_list, 'maximum_units' ), SORT_ASC, $price_tier_list );
```

File: projects/packages/search/src/class-search-products.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace Automattic\Jetpack\Search;
+
+/**
+ * Class to fetch Search product pricing
+ *
+ * @package Automattic\Jetpack\Search
+ */
+class Search_Products {
+	const DEFAULT_PROMOTED_PRODCT = 'jetpack_search';
```
Comment (bluefuton) [inline projects/packages/search/src/class-search-products.php:16]:
```suggestion
	const DEFAULT_PROMOTED_PRODUCT = 'jetpack_search';
```

File: projects/packages/search/src/class-search-products.php [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		$products        = json_decode( wp_remote_retrieve_body( $wpcom_request ), true );
+		$search_products = array_filter(
+			$products,
+			function ( $product, $key ) {
+				return 0 === strpos( 'jetpack_search', $key );
+			},
+			ARRAY_FILTER_USE_BOTH
+		);
+		// We don't want to cache error response for too long.
+		$time_to_cache = ! empty( $search_products ) ? DAY_IN_SECONDS : 5 * MIMUTES_IN_SECONDS;
```
Comment (bluefuton) [inline projects/packages/search/src/class-search-products.php:86]:
```suggestion
		$time_to_cache = ! empty( $search_products ) ? DAY_IN_SECONDS : 5 * MINUTES_IN_SECONDS;
```

File: projects/packages/search/src/class-search-products.php [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+namespace Automattic\Jetpack\Search;
+
+/**
+ * Class to fetch Search product pricing
+ *
+ * @package Automattic\Jetpack\Search
+ */
+class Search_Products {
```
Comment (bluefuton) [inline projects/packages/search/src/class-search-products.php:15]: Not sure we have a Jetpack convention on this, but have we used singular naming for other classes (so `Search_Product`)?

Comment (kangzj) [reply projects/packages/search/src/class-search-products.php]: Good question, renamed to `Product` 😄 a97f135

File: projects/packages/search/src/class-search-products.php (hunk 3 above)
Comment (kangzj) [reply projects/packages/search/src/class-search-products.php]: addressed in a97f135

File: projects/packages/search/src/class-search-products.php (hunk 2 above)
Comment (kangzj) [reply projects/packages/search/src/class-search-products.php]: addressed in a97f135

File: projects/packages/search/src/class-search-products.php (hunk 1 above)
Comment (kangzj) [reply projects/packages/search/src/class-search-products.php]: addressed in a97f135

**Review by jsnmoon — DISMISSED** [review]
Works well with the corresponding patch.

**Review by jsnmoon — DISMISSED** [review]

**Review by jsnmoon — APPROVED** [review]

**Pull request merged**